q2 plugin

mortonjt opened the pull request:
Here's the prototype for the qiime2 plugin.

Note that this also depends on #113 .

mortonjt commented on gneiss/q2/_type.py:
```diff
@@ -0,0 +1,23 @@
+
+# Tree types
+Hierarchy = SemanticType('Hierarchy', field_names=['type'])
+Cluster = SemanticType('Cluster', variant_of=Hierarchy.field['type'])
```
@jairideout @gregcaporaso this was the original motivation for raising https://github.com/qiime2/q2-types/issues/99

While there are some scenarios that can use a phylogeny for an analysis, most scenarios will deal with different types of hierarchies, whether it be generated from a clustering or something else.

I see a few options for this

1. Define a completely orthogonal type to phylogeny named `Hierarchy`
2. Make `Hierarchy` a super type

Also, would we have to worry about clashing types?  For example, if I have defined `Hierarchy[Cluster]` here, could that potentially clash with an unsupervised sampling clustering type implemented in a classifier package?

jairideout replied:
cc @ebolyen

> Also, would we have to worry about clashing types? For example, if I have defined Hierarchy[Cluster] here, could that potentially clash with an unsupervised sampling clustering type implemented in a classifier package?

Types would only clash if the same name(s) were registered by those two plugins. It might be worth thinking about what names make sense for a classifier plugin so we can come up with distinct names (or reuse the types?).

ebolyen replied:
Just to make sure that this is known, I'd like to point out  that you can define ad-hoc super-types with the `|` operator. e.g. `Phylogony[Rooted] | HierarchicalCluster | Graph`, etc. This might be the best way to go for now, but it doesn't help us with naming the other pieces. 

I like the idea of `Hiearchy`, but what other kinds of variant-types would we have besides `Cluster`? Ideally this could be reused for a classifier, as long as the data was strictly a hierarchy of clusters and not much else.

mortonjt replied:
In context of the balances, there are quite a few more types.

The tree could be produced to represent a gradient along an environment.
It could also be used to represent the largest variance partitions, analogous to Principle Components Analysis.  We have a few more techniques in the works that will yield even more different ways to generate trees specific for features.

For samples, there is hierarchical clustering, but there is also decision trees.  I'd think it would be a good idea to keep trees relating samples and trees relating features to be distinct entities, but maybe stem from the same super type.

mortonjt commented on gneiss/q2/_type.py:
```diff
@@ -0,0 +1,23 @@
+
+
+# Regression types
+Regression = SemanticType('Regression', field_names=['type'])
```
Thinking about changing this to `BalanceRegression` to avoid potential name clashing with other Regression types.  For example there are techniques such as Multinomial Regression or Distance based Regression

mortonjt commented:
@RNAer have you seen this error before?  It looks like the conda binary cannot be retrieved?

RNAer commented:
try to restart the travis

mortonjt commented:
I did - still doesn't work ...

RNAer commented:
what do you mean? It clearly show you have pep8 error

````
pep8 gneiss setup.py

gneiss/util.py:192:1: E302 expected 2 blank lines, found 1

gneiss/regression/_ols.py:31:1: E302 expected 2 blank lines, found 1

gneiss/regression/tests/test_mixedlm.py:61:1: E302 expected 2 blank lines, found 1

gneiss/regression/tests/test_ols.py:54:5: E303 too many blank lines (2)

make: *** [pep8] Error 1
```

wasade commented on gneiss/regression/_mixedlm.py:
```diff
@@ -6,13 +6,24 @@
 from decimal import Decimal
 from statsmodels.iolib.summary2 import Summary
+from gneiss.util import HAVE_Q2
+if HAVE_Q2:
```
why not just make gneiss depend on q2?

wasade commented on gneiss/regression/_mixedlm.py:
```diff
@@ -260,3 +271,27 @@ def _format(x):
+    return res
+
+
+plugin.methods.register_function(
```
...the downside to not just depending on q2 is that bugs like this can slip through very easily. `plugin` is a `NameError` here.

wasade commented on gneiss/regression/_model.py:
```diff
@@ -186,3 +191,25 @@ def pvalues(self):
+    class RegressionFormat_g(model.BinaryFileFormat):
+        def sniff(self):
+            try:
+                res = RegressionModel.read_pickle(open(str(self)))
```
the sniff is a full parse?

wasade commented on gneiss/util.py:
```diff
@@ -184,3 +195,81 @@ def rename_internal_nodes(tree, names=None, inplace=False):
     return _tree
+
+
+def _intersect_of_table_metadata_tree(table, metadata, tree):
```
Did this commit sneak in? Doesn't seem like its part of the plugin?

mortonjt replied:
This is part of refactoring the code.  After evaluating where this sort of functionality really should belong, I think it is more appropriate that it should go into `gneiss.util` since it is a utility function.

wasade commented on setup.py:
```diff
@@ -88,7 +88,13 @@ def finalize_options(self):
           'ete3',
       ],
+      extra_require={
+          'q2': ['qiime >= 2.0.5', 'biom-format', 'seaborn']
```
should this be `2017.2.0`?

mortonjt commented:
The q2 interface how appears to be working (thanks to @ebolyen, @jakereps and @thermokarst)!

@wasade @sjanssen2 will you be able to review this?  Thanks!!

sjanssen2 commented on gneiss/regression/_format.py:
```diff
@@ -0,0 +1,31 @@
+
+class RegressionFormat_g(model.BinaryFileFormat):
+    def sniff(self):
+        try:
```
you also might want to check the type of the pickled object to have a little bit of contents check

mortonjt replied:
Done

sjanssen2 commented on gneiss/regression/_regression.py:
```diff
@@ -5,86 +5,54 @@
-                             columns=non_tips,
-                             index=table.index)
-    return ilr_table, basis
+import skbio
```
isn't that too general? But I don't know how much overhead time it takes to load all of skbio instead of more specific modules.

mortonjt replied:
I don't think that will be a problem.  `pandas` and `numpy` are much larger projects than `skbio`, but we regularly load all of the modules from those libraries

sjanssen2 commented on gneiss/regression/_regression.py:
```diff
@@ -5,86 +5,54 @@
+from ._type import Regression_g, Linear_g, LinearMixedEffects_g
+
+
+def ols_regression(table: pd.DataFrame, tree: skbio.TreeNode,
```
cool, a signature like in Haskell :-)

sjanssen2 replied:
even if this method is trivial, shouldn't you provide a docstring? Or at least a TODO that the docstring needs to be written in the future?

wasade replied:
@sjanssen2 it's not really necessary for the qiime2 plugin methods as help text and docs based off the parameters are handled by the plugin API itself based off the info provided at registration

sjanssen2 commented on gneiss/regression/_transformer.py:
```diff
@@ -0,0 +1,33 @@
+    ff = RegressionFormat_g()
+    with ff.open() as fh:
+        data.write_pickle(fh)
+    return ff
```
is it ensured that the filehandle is closed? I recently found that there is a problem in BIOM.

wasade replied:
Can you open an [issue](https://github.com/biocore/biom-format/issues) with BIOM please? 

In this case, the file handle (`fh`) is closed by the context manager.

mortonjt replied:
Not sure what is going on - but it seems to appear in quite a few projects

https://github.com/scikit-learn/scikit-learn/issues/3410
https://github.com/python-pillow/Pillow/issues/477
https://github.com/Supervisor/supervisor/issues/391

sjanssen2 commented on gneiss/regression/_mixedlm.py:
```diff
@@ -9,8 +9,8 @@
+                         _to_balances)
 from decimal import Decimal
 from statsmodels.iolib.summary2 import Summary
 
```
line 142 ff are not covered by unittest. Could you please add a test that checks that the ValueError is raised

mortonjt replied:
Done

sjanssen2 commented on gneiss/regression/tests/test_format.py:
```diff
@@ -0,0 +1,42 @@
+        format = RegressionFormat_g(filepath, mode='r')
+
+        print(filepath, str(format), format.sniff())
+        format.validate()
```
I still new to unittests in Python. Don't you have to add some kind of self.assertxxx statements to the code?

wasade replied:
:+1:

wasade replied:
...although, I think `validate` just returns `None` if valid and raises if otherwise

sjanssen2 commented on gneiss/regression/tests/test_regression.py:
```diff
@@ -0,0 +1,118 @@
+        proc.wait()
+
+        self.assertTrue(os.path.exists("test_ols.qza"))
+
```
can you also add a test to compare the file contents? Maybe a call to a unix diff?

mortonjt replied:
Adding in @wasade 's suggestion

sjanssen2 commented on gneiss/regression/tests/test_regression.py:
```diff
@@ -0,0 +1,118 @@
+
+        self.assertTrue(os.path.exists("test_lme.qza"))
+        os.remove("test_lme.qza")
+
```
see above about using diff

mortonjt replied:
This has been addressed using the Artifact API.  Really want to avoid using tar here.

sjanssen2 commented on gneiss/util.py:
```diff
@@ -184,3 +188,81 @@ def rename_internal_nodes(tree, names=None, inplace=False):
+        correspond to columns.
+    tree : skbio.TreeNode
+        Tree object where the leaves correspond to the columns contained in
+        the table.
```
do you mean "the object table" or the metadata-table by "table"?

mortonjt replied:
the table containing all of the samples.

sjanssen2 commented on gneiss/util.py:
```diff
@@ -184,3 +188,81 @@ def rename_internal_nodes(tree, names=None, inplace=False):
+    pd.DataFrame
+        Subset of `metadata` with common row names as `table`
+    skbio.TreeNode
+        Subtree of `tree` with common tips as `table`
```
Is it a true sub-tree or a tree where some tips and according branches are removed?

wasade commented on gneiss/regression/_format.py:
```diff
@@ -0,0 +1,31 @@
+        try:
+            # just check if the file is pickleable.
+            with open(str(self), 'rb') as fh:
+                pickle.load(fh)
```
if the pickle is large or complex, this will be an expensive sniff

mortonjt replied:
Yeah - I couldn't think of a better way to do it, since pickle files don't have headers. 
Do you have a better idea?

mortonjt replied:
Also keep in mind that there is an astronomical development burden on reinventing another file format specific for storing Regression objects - so pickle is a strict requirement for the time being.

wasade commented on gneiss/regression/_format.py:
```diff
@@ -0,0 +1,31 @@
+                pickle.load(fh)
+            return True
+        except:
+            return None
```
shouldn't this be `False`?

mortonjt replied:
I don't think so

Looking at `NewickFormat`, `get_sniffer` is called.
[get_sniffer](https://github.com/biocore/scikit-bio/blob/master/skbio/io/registry.py#L248) returns `None` if there is no file format detected.

wasade commented on gneiss/regression/_regression.py:
```diff
@@ -5,86 +5,54 @@
+    function=ols_regression,
+    inputs={'table': FeatureTable[Composition],
+            'tree': Phylogeny[Rooted | Unrooted]},
+    parameters={'formula': Str, 'metadata': Metadata},
```
Can you add parameter and output descriptions? For example, see [here](https://github.com/qiime2/q2-diversity/blob/master/q2_diversity/plugin_setup.py#L35-L47).

mortonjt replied:
Done

wasade commented on gneiss/regression/_regression.py:
```diff
@@ -5,86 +5,54 @@
+
+plugin.methods.register_function(
+    function=lme_regression,
+    inputs={'table': FeatureTable[Composition],
```
same comment re: descriptions

wasade commented on gneiss/regression/_transformer.py:
```diff
@@ -0,0 +1,33 @@
+@plugin.register_transformer
+def _2(ff: RegressionFormat_g) -> OLSModel:
+    with ff.open() as fh:
+        print(fh)
```
Remove please

wasade commented on gneiss/regression/_transformer.py:
```diff
@@ -0,0 +1,33 @@
+    return ff
+
+
+@plugin.register_transformer
```
Would it make sense to have a transform from `LMEModel` -> `OLSModel`? It is possible now indirectly via `RegressionFormat_g` but that doesn't mean it's something that should be done...

mortonjt replied:
I don't think it would make sense to have such a transform, since they are two very distinct objects.  Given this, do you think it is a better idea to create two formats for each of these models?

wasade commented on gneiss/regression/tests/data/not-regression.pickle:
```diff
@@ -0,0 +1 @@
+asdfasdfasdf
```
is this a valid `.pickle` file?

mortonjt replied:
Nope - pickle files are binaries.

wasade commented on gneiss/regression/tests/test_format.py:
```diff
@@ -0,0 +1,42 @@
+        filepath = self.get_data_path('ols.pickle')
+        format = RegressionFormat_g(filepath, mode='r')
+
+        print(filepath, str(format), format.sniff())
```
Remove please

wasade commented on gneiss/regression/tests/test_regression.py:
```diff
@@ -0,0 +1,118 @@
+        pdt.assert_frame_equal(exp_resid, res.residuals())
+
+    def test_ols_cli(self):
+        # TODO: Is there a better way to test this?
```
Yes, just load the artifacts and pass them in:

```python
from qiime2.plugins.gneiss.methods import ols_regression

in_table = qiime2.Artifact.load('path/to/foo.qza')
in_tree = qiime2.Artifact.load('path/to/foo.qza')
in_metadata = qiime2.Artifact.load('path/to/foo.qza')

results = ols_regression(table=in_table, tree=in_tree, metadata=in_metadata, etc etc)
```

Note that is a `Results` object. It is a `namedtuple` so you can fetch the individual returned artifacts out easily. If it's just a single artifact coming back, you can implicitly unpack it with:

```python
result, = ols_regression(table=in_table, tree=in_tree, metadata=in_metadata, etc etc)
```

...in which case, `result` would be an artifact of the semantic type as returned by the API method based on the plugin registration.

mortonjt replied:
Thanks!

wasade commented on gneiss/regression/tests/test_regression.py:
```diff
@@ -0,0 +1,118 @@
+                               check_less_precise=True)
+
+    def test_lme_cli(self):
+        # TODO: Is there a better way to test this?
```
see above

wasade commented on gneiss/util.py:
```diff
@@ -184,3 +188,81 @@ def rename_internal_nodes(tree, names=None, inplace=False):
+
+
+def _intersect_of_table_metadata_tree(table, metadata, tree):
+    """ Matches tips, features and samples between the table, metadata
```
this should be a single sentence for the first line of the doc string

mortonjt replied:
I don't think I can get this to fit within 1 line.  Do you think it is really necessary for a helper method?

wasade replied:
What about "The intersection of tips and features" and then the more specific detail follows? I believe samples and metadata subset are implicit as driven by the overlap in tips and features, right?

wasade commented on gneiss/util.py:
```diff
@@ -184,3 +188,81 @@ def rename_internal_nodes(tree, names=None, inplace=False):
+                         'Use pseudocounts or ``multiplicative_replacement``.'
+                         )
+    # check to see if there are overlapping nodes in tree and table
+    overlap = {n.name for n in tree.traverse()} & set(table.columns)
```
shouldn't this be `tree.tips()`?

mortonjt replied:
Yup - the only problem is we specifically need strings for the names rather than the entire `TreeNode`.  So we'll still need the set comprehension in the end.

mortonjt commented:
The bulk of the comments have been addressed.

Now I have made some large changes here, namely

1. Ditched the `Regression_g` super type and changed `Linear_g` -> `LinearRegression_g`.
2. Split up the `RegressionFormat_g` into two different formats, namely `LinearRegressionFormat_g` and `LinearMixedEffectsFormat_g`.  This ultimately allowed for more fine-tuned checks for the underlying models

This ultimately improves the robustness of these objects and avoids overgeneralizing.

@wasade @sjanssen2 do you mind if you could do one more sanity check once the tests pass? 
We can merge this once you guys think this is ok.  Thanks!

wasade commented on gneiss/regression/tests/test_regression.py:
```diff
@@ -0,0 +1,154 @@
+                                 columns=['Y1', 'Y2'])
+        pdt.assert_frame_equal(exp_resid, res.residuals())
+
+    def test_ols_cli(self):
```
I think this is redundant with `test_ols_artifact`?

mortonjt replied:
I think it is ok to keep those tests in, just as a sanity check that the Artifact API / CLI are working as expected.  This is mainly to serve as an alert, especially since the q2 repo is currently in flux.

mortonjt commented:
🎆  tests passing!
